### PR TITLE
add vector dig in associable module

### PR DESCRIPTION
**Pull request by HParker**
part of #212

pull out hash dig and use it in the `associable` module.

### CODE REVIEW

File: lib/hamster/associable.rb [hunk 1]
```diff
@@ -45,5 +45,23 @@ def update_in(*key_path, &block)
       end
       put(key, new_value)
     end
+
+    # Return the value of successively indexing into a hash
```
Comment (alexdowad) [inline lib/hamster/associable.rb]: "successively indexing into a *hash*"...

**Comment by HParker** [comment]
Cool, I think the documentation makes more sense for Vector and Hash.

**Comment by alexdowad** [comment]
Any comments from other contributors?

File: lib/hamster/associable.rb [hunk 2]
```diff
@@ -45,5 +45,28 @@ def update_in(*key_path, &block)
       end
       put(key, new_value)
     end
+
+    # Return the value of successively indexing into a container
+    # If any of the keys is not present in the container, return `nil`.
+    # keys that the Hamster type doesn't understand, raises an argument error
+    #
+    # @example
+    #   h = Hamster::Hash[:a => 9, :b => Hamster::Hash[:c => 'a', :d => 4], :e => nil]
```
Comment (dubek) [inline lib/hamster/associable.rb]: I would give an example that has a Hash+Vector together, like the first example of `#update_in` above. This shows that both containers can be used together with the same method.

File: lib/hamster/associable.rb [hunk 3]
```diff
@@ -45,5 +45,28 @@ def update_in(*key_path, &block)
       end
       put(key, new_value)
     end
+
+    # Return the value of successively indexing into a container
```
Comment (dubek) [inline lib/hamster/associable.rb]: Add a period at the end of the first sentence.

**Comment by dubek** [comment]
Besides my small code comments on the documentation, I just noticed this at the top of `associable.rb`:

```
   # To mix in `Associable`, your class must implement two methods:
   #
   # * `fetch(index, default = (missing_default = true))`
   # * `put(index, item = yield(get(index)))`
   #
```

However the new `#dig` method uses `[]` and not `fetch`. Can we change to `fetch` without breaking or making it over-complex?

File: lib/hamster/associable.rb (hunk 3 above)
Comment (alexdowad) [inline lib/hamster/associable.rb]: I would add that "collection" is preferable to "container".

**Comment by HParker** [comment]
@dubek, it totally works with `fetch` only issue is we have to rescue `IndexError`s since indexing off the end of an array with `dig` returns `nil`, but `fetch` raises an error.

Let me know if you prefer this, and I'll squish the commits.

**Comment by HParker** [comment]
ah, sorry disregard my previous comment. I just had to stare at the code for a minute longer. I always forget about `fetch` default.

**Comment by dubek** [comment]
Thanks for the quick fixes.

I just thought about the case of Hash with default proc. In Ruby 2.3.0-preview1 it behaves as follows:

```
2.3.0-preview1 :001 > h = Hash.new { |h,k| "#{k}-default-val" }
 => {} 
2.3.0-preview1 :002 > h[:a]
 => "a-default-val" 
2.3.0-preview1 :003 > h.fetch(:a)
KeyError: key not found: :a
	from (irb):3:in `fetch'
	from (irb):3
	from /home/dmurik/.rvm/rubies/ruby-2.3.0-preview1/bin/irb:11:in `<main>'
2.3.0-preview1 :004 > h.fetch(:a, nil)
 => nil 
2.3.0-preview1 :005 > h.dig(:a)
 => "a-default-val" 
```

@HParker - Can you please check where the suggest `#dig` works similarly? Or will we need to resort to using `[]` as you originally suggested?

In any case we need to make sure the `Hamster::Hash#dig` specs indeed specify these cases (with and without default proc).

**Comment by HParker** [comment]
Thanks for all the great feedback.
good catch. I hadn't thought of that case.
I can't think of a good way of implementing this with `fetch`. How would you feel about either another module with a dependency on `get` or adding a dependency on `get` to this module?

Do you have any other ideas?

**Comment by HParker** [comment]
I came across another option. ruby hash's expose a `default` on hashes. If we had a similar method on hamster hashes, we could use that to get the correct default rather than assuming nil.

**Comment by alexdowad** [comment]
> I can't think of a good way of implementing this with fetch

If so, adding a dependency on `get` is not really a problem.

**Comment by dubek** [comment]
Yep, I agree. Just use `#get` and add it to the documentation of `Associable` as a required method.

**Comment by HParker** [comment]
cool, we are back to the `get` implementation. Thanks for all the help everyone. Seriously good contribution experience.

**Comment by alexdowad** [comment]
I think this is ready to merge.

**Comment by dubek** [comment]
Rebased and merged to master manually.

Thanks @HParker for dealing with all the back and forth! Great contribution.

**Pull request closed without merging**